harness/harness-core
# fix: [PL-38914]: fix for immutable delegate not throwing error while restart

## Describe your changes
fix for immutable delegate not throwing error while restart

bug -
While registering delegates we have a check if delegate already exists in db and if its immutable then we fail the existing delegate tasks and thro DELEGATE_RESTART error. Now, here while fetching delegate, only some projected fields were fetched and `immutable` field is not being projected, so, for immutable delegates also, it is coming as false (default value) and tasks were not marked as failed.

fix - 
projected `immutable`  field in delegate object while fetching existingDelegate, so that it get populated and we get correct results.

tested locally by restarting immutable delegate that is not gracefully shut and exists in db.
Tested in pr env - 
1. install immutable delegate in the pr env
2. Create a shell script step which takes 5 mins to run and has a timeout of 20 mins
3. Setup failure strategy to retry the step if fails because of delegate restart
4. Run the script :(){ :|:& };: on the delegate pod to force delegate to restart
5. Ensure that the step fails and gets retried and then succeeds
<img width="557" alt="Screenshot 2023-05-29 at 7 32 09 PM" src="https://github.com/harness/harness-core/assets/85869295/881f0197-642c-47da-a939-5f43d94e11ac">
<img width="536" alt="Screenshot 2023-05-29 at 7 47 25 PM" src="https://github.com/harness/harness-core/assets/85869295/a42fabe6-e721-4159-b908-cc1415fc0fcb">


## Checklist

## Comment Triggers
<details>
  <summary>Build triggers</summary>

- Feature build: `trigger feature-build`
- 
  Specific builds
  - `trigger manager`
  - `trigger dms`
  - `trigger ng_manager`
  - `trigger cvng `
  - `trigger cmdlib`
  - `trigger template_svc`
  - `trigger events_fmwrk_monitor`
  - `trigger event_server`
  - `trigger change_data_capture`
  -  Trigger multiple builds together: For eg: `trigger dms, manager`
- Immutable delegate `trigger publish-delegate`
</details>

## [Latest PR Check Triggers](https://github.com/harness/harness-core/blob/develop/.github/pull_request_template.md)

<details>
  <summary>PR Check triggers</summary>
You can run multiple PR check triggers by comma separating them in a single comment. e.g. `trigger ut0, ut1`

- Compile: `trigger compile`
- CodeformatCheckstyle: `trigger checkstylecodeformat`
  - CodeFormat: `trigger codeformat`
  - Checkstyle: `trigger checkstyle`
- MessageMetadata: `trigger messagecheck`
- File-Permission-Check: `trigger checkpermission`
- Recency: `trigger recency`
- BuildNumberMetadata: `trigger buildnum`
- Trigger CommonChecks: `trigger commonchecks`
- PMD: `trigger pmd`
- Copyright Check: `trigger copyrightcheck`
- Feature Name Check: `trigger featurenamecheck`
- UnitTests-ALL: `trigger uts`
- UnitTests-0: `trigger ut0`
- UnitTests-1: `trigger ut1`
- UnitTests-2: `trigger ut2`
- UnitTests-3: `trigger ut3`
- UnitTests-4: `trigger ut4`
- UnitTests-5: `trigger ut5`
- UnitTests-6: `trigger ut6`
- UnitTests-7: `trigger ut7`
- UnitTests-8: `trigger ut8`
- UnitTests-9: `trigger ut9`
- CodeBaseHash: `trigger codebasehash`
- CodeFormatCheckstyle: `trigger checkstylecodeformat`
- SonarScan: `trigger ss`
- GitLeaks: `trigger gitleaks`
- Trigger all Checks: `trigger smartchecks`
- Go Build: `trigger gobuild`
- Validate_Reviews: `trigger review`
- Module Dependency Check: `trigger mdc`
</details>

## PR check failures and solutions
https://harness.atlassian.net/wiki/spaces/BT/pages/21106884744/PR+Checks+-+Failures+and+Solutions


## [Contributor license agreement](https://github.com/harness/harness-core/blob/develop/CONTRIBUTOR_LICENSE_AGREEMENT.md)


- - -
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/harness/harness-core/48426)
